configure.ac

```diff
@@ -578,7 +578,7 @@ AS_CASE([$host],
 -fexcess-precision=standard"
         # TODO: see whether the code can be fixed to avoid -Wno-unused
         common_cflags="-O2 -fno-strict-aliasing -fwrapv -mms-bitfields"
-        internal_cppflags='-DUNICODE -D_UNICODE'
+        internal_cppflags='-D__USE_MINGW_ANSI_STDIO=0 -DUNICODE -D_UNICODE'
```

Comment:
Naive question: why do we need both this define on the command line and the `#undef`/`#define` dance in `caml/config.h`?

Reply:
This one is for the build (so our calls to the C compiler) - the dance in `config.h` is for the installation (which includes `-output-complete-obj` and any code which uses `misc.h` with `CAML_INTERNALS` defined)

Reply:
OK for the "installation" part, but I'm still curious: if we did not have `-D__USE_MINGW_ANSI_STDIO=0` in the flags used for the build, would the `config.h` dance do the right thing nonetheless?

Reply:
Hah - that has me curious now! Two seconds...

Reply:
OK, there were other tests running so that was a long two seconds. I just checked, and I remember it now: for the runtime, we really do need to disable the mingw-w64 C99 extensions completely, because it affects printf formats (which remain the older-style for older MSVC compatibility). So the runtime _needs_ `__USE_MINGW_ANSI_STDIO=0` to disable its extended version of `printf`.

For external uses where `caml/misc.h` is included **and** `CAML_INTERNALS` is defined (so caveat emptor, big time) the problem is completely different and just about symbol redefinition. The issue here is that mingw's `stdio.h` _always_ defines a function `snprintf` (and always has). When `__USE_MINGW_ANSI_STDIO` is `0`, it politely `#undef`'s any existing `snprintf` macro before doing this (but using cpp's push/pop mechanism) but it doesn't do this (inconsistently) if `__USE_MINGW_ANSI_STDIO` is `1`. So after `misc.h` is read, we need to be sure that if `__USE_MINGW_ANSI_STDIO` is `0` (and the user has to accept this because they defined `CAML_INTERNALS`) because otherwise any subsequent reading of `stdio.h` will fail (which is where the original error report came from) since it will try to define a function `caml_snprintf` a second time because the `snprintf` macro will not have been `#undef`'d.

Reply:
OK, now I understand what's going on.  Thanks for the explanations.